group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -0,0 +1,535 @@
+                updateSubscription.accept(member.subscribedTopicNames());
+            }
+        });
+
+        return Collections.unmodifiableMap(newSubscriptionMetadata);
+    }
+
+    /**
+     * Updates the current state of the group.
+     */
+    private void maybeUpdateGroupState() {
+        if (members.isEmpty()) {
+            state.set(ConsumerGroupState.EMPTY);
+        } else if (groupEpoch.get() > assignmentEpoch.get()) {
+            state.set(ConsumerGroupState.ASSIGNING);
+        } else {
+            for (Map.Entry<String, ConsumerGroupMember> keyValue : members.entrySet()) {
+                ConsumerGroupMember member = keyValue.getValue();
+                if (member.nextMemberEpoch() != assignmentEpoch.get() || member.state() != ConsumerGroupMember.MemberState.STABLE) {
+                    state.set(ConsumerGroupState.RECONCILING);
+                    return;
+                }
+            }
+
+            state.set(ConsumerGroupState.STABLE);
+        }
+    }
+
+    /**
+     * Updates the partition epochs based on the old and the new member.
+     */
+    private void maybeUpdatePartitionEpoch(
+        ConsumerGroupMember oldMember,
+        ConsumerGroupMember newMember
+    ) {
+        if (oldMember == null) {
+            addPartitionEpochs(newMember.assignedPartitions(), newMember.memberEpoch());
+            addPartitionEpochs(newMember.partitionsPendingRevocation(), newMember.memberEpoch());
+        } else {
+            if (!oldMember.assignedPartitions().equals(newMember.assignedPartitions())) {
+                removePartitionEpochs(oldMember.assignedPartitions());
+                addPartitionEpochs(newMember.assignedPartitions(), newMember.memberEpoch());
+            }
+            if (!oldMember.partitionsPendingRevocation().equals(newMember.partitionsPendingRevocation())) {
+                removePartitionEpochs(oldMember.partitionsPendingRevocation());
+                addPartitionEpochs(newMember.partitionsPendingRevocation(), newMember.memberEpoch());
+            }
+        }
+    }
+
+    /**
+     * Removes the partition epochs for the provided member.
+     */
+    private void maybeRemovePartitionEpoch(
+        ConsumerGroupMember oldMember
+    ) {
+        if (oldMember != null) {
+            removePartitionEpochs(oldMember.assignedPartitions());
+            removePartitionEpochs(oldMember.partitionsPendingRevocation());
+        }
+    }
+
+    /**
+     * Removes the partition epochs based on the provided assignment.
+     */
+    private void removePartitionEpochs(
+        Map<Uuid, Set<Integer>> assignment
+    ) {
+        assignment.forEach((topicId, assignedPartitions) -> {
+            currentPartitionEpoch.compute(topicId, (__, partitionsOrNull) -> {
+                if (partitionsOrNull != null) {
+                    assignedPartitions.forEach(partitionsOrNull::remove);
+                    if (partitionsOrNull.isEmpty()) {
+                        return null;
```

Comment:
I thought the java doc for TimelineHashMap doesn't allow null values.

Reply:
Right but `compute` gives you `null` if the key is not present.

Reply:
if we compute and return null, isn't that trying to set null values? Or am I missing something.

Reply:
I guess I'm also wondering why we don't use empty set. Unless this is because we want to overload null as mentioned in the getEpoch method.

Reply:
From the javadoc of `compute`:

```
If the remapping function returns null, the mapping is removed (or remains absent if initially absent). If the remapping function itself throws an (unchecked) exception, the exception is rethrown, and the current mapping is left unchanged.
```

We don't keep the empty set because we want to clean the Map. Otherwise, it would keep sets for non-existing partitions for instance.

Reply:
Clarified offline -- null returned in compute removes the entry.